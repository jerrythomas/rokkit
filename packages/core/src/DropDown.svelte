<script>
	import Icon from './Icon.svelte'
	import Text from './items/Text.svelte'
	import ListItems from './ListItems.svelte'
	import { defaultFields } from './constants.js'
	import { dismissable } from './actions'
	import { createEventDispatcher } from 'svelte'

	const dispatch = createEventDispatcher()

	let className = ''
	export { className as class }
	export let items = []
	export let fields = defaultFields
	export let using = { default: Text }
	export let value = null
	export let placeholder = 'Select a value'

	$: using = { default: Text, ...using }
	$: fields = { ...defaultFields, ...fields }

	let offsetTop = 0
	let open = false

	function handleSelect(event) {
		open = false
		dispatch('change', event.detail)
	}
</script>

<!-- svelte-ignore a11y-no-noninteractive-tabindex -->
<div
	class="flex w-full relative cursor-pointer dropdown {className}"
	class:open
	tabindex="0"
	use:dismissable
	on:blur={() => (open = false)}
	on:dismiss={() => (open = false)}
>
	<button
		on:click|stopPropagation={() => (open = !open)}
		class="flex w-full"
		bind:clientHeight={offsetTop}
		tabindex="-1"
	>
		<svelte:component this={using.default} content={value ?? placeholder} />
		<Icon name="dropdown-opened" />
	</button>
	{#if open}
		<div
			class="flex flex-col absolute z-10 h-fit w-full menu"
			style:top="{offsetTop}px"
		>
			<ListItems
				{items}
				{fields}
				{using}
				bind:activeItem={value}
				on:select={handleSelect}
			/>
		</div>
	{/if}
</div>
