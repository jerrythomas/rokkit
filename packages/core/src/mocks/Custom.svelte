<script>
	export let value
	export let fields
</script>

<span>{value[fields.text]}</span>
