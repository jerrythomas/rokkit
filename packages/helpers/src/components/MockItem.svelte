<script>
	let { value } = $props()
</script>

{value}
