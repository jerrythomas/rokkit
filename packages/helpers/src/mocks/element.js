import { vi } from 'vitest'

/**
 * Creates an array of elements with the specified size
 *
 * @param {number} count
 * @param {number} size
 * @param {string} [prop='offsetHeight']
 * @returns {Array<Object<string, number>>}
 */
export function elementsWithSize(count, size, prop = 'offsetHeight') {
	return Array.from({ length: count }, () => ({
		[prop]: size
	}))
}

/**
 * Creates an array of elements with mixed sizes
 *
 * @param {Array<{count: number, size: number}>} data
 * @param {string} prop
 * @returns {Array<Object<string, number>>}
 */
export function mixedSizeElements(data, prop) {
	return data.reduce(
		(elements, { count, size }) => [...elements, ...elementsWithSize(count, size, prop)],
		[]
	)
}

/**
 * Creates a mock node with functions to add and remove event handlers
 *
 * @param {Array<string} events
 * @returns {{node: HTMLElement, listeners: Object<string, integer>}}
 */
export function getMockNode(events) {
	const listeners = events.reduce((acc, event) => ({ ...acc, [event]: 0 }), {})

	const node = {
		dispatchEvent: vi.fn(),
		scrollTo: vi.fn(),
		querySelectorAll: vi.fn().mockImplementation((selector) => {
			return [document.createElement(selector)]
		}),
		querySelector: vi.fn().mockImplementation((selector) => {
			return document.createElement(selector)
		}),
		addEventListener: vi.fn().mockImplementation((name) => ++listeners[name]),
		removeEventListener: vi.fn().mockImplementation((name) => --listeners[name])
	}
	return { node, listeners }
}

/**
 * @typedef {Object} NestedItem
 * @property {string} name
 * @property {string} [dataPath]
 * @property {string} [id]
 * @property {Array<NestedItem>} [children]
 */

/**
 * Creates a nested HTML element structure using the provided data
 *
 * @param {NestedItem} item
 * @returns {HTMLElement}
 */
export function createNestedElement(item) {
	const { name, dataPath, id, children } = item
	const element = document.createElement(name)

	if (dataPath) element.dataset.path = dataPath
	if (id) element.id = id

	element.scrollIntoView = vi.fn()

	if (Array.isArray(children)) {
		children.forEach((child) => element.appendChild(createNestedElement(child)))
	}

	return element
}

/**
 * Mocks the requestSubmit function for HTMLFormElement which is not implemented by JSDOM
 */
export function mockFormRequestSubmit() {
	// eslint-disable-next-line no-undef
	HTMLFormElement.prototype.requestSubmit = vi.fn(function (submitter) {
		const event = new Event('submit', {
			bubbles: true,
			cancelable: true
		})
		if (submitter) {
			event.submitter = submitter
		}
		this.dispatchEvent(event)
	})

	return true
}
