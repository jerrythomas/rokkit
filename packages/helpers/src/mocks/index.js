import './animate'
import { ResizeObserver } from './resize-observer'

// skipcq: JS-E1004 - Needed for exposing all functions
export * from './match-media'
// skipcq: JS-E1004 - Needed for exposing all functions
export * from './element'

global.ResizeObserver = ResizeObserver
