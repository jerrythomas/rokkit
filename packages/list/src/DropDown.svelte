<script>
	import List from './List.svelte'
	import ListItem from './ListItem.svelte'
	import Slider from './Slider.svelte'
	import DropDownIcon from './icons/DropDownIcon.svelte'

	export let data
	export let component = null
	export let selected = null
	export let key = 'id'
	export let selectedKey

	let opened = false

	function handleSelect(event) {
		console.log('selected', event)
		selected = event.detail
	}
</script>

<div
	class="flex flex-col outline-none w-full h-12 relative z-1 dropdown"
	class:opened
	on:focus={() => (opened = true)}
	on:blur={() => (opened = false)}
	tabindex={0}
>
	<div class="flex flex-shrink-0 h-12 items-center pl-4 selected-item">
		<span class="flex flex-grow">
			<ListItem {component} item={selected} />
		</span>
		<DropDownIcon />
	</div>

	{#if opened}
		<Slider class="h-60 absolute top-12 w-full z-1">
			<List
				bind:items={data}
				{key}
				{component}
				bind:selected={selectedKey}
				on:select={handleSelect}
				on:change
			/>
		</Slider>
	{/if}
</div>
