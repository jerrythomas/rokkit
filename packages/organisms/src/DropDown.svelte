<script>
	import { defaultFields, defaultStateIcons, createEmitter } from '@rokkit/core'
	import { dismissable } from '@rokkit/actions'
	import { Icon, Slider } from '@rokkit/atoms'
	import { Item } from '@rokkit/molecules'
	import List from './List.svelte'

	/**
	 * @typedef {Object} Props
	 * @property {string} [class]
	 * @property {any} [options]
	 * @property {import('@rokkit/core').FieldMapping} [fields]
	 * @property {any} [using]
	 * @property {any} [value]
	 * @property {any} [title]
	 * @property {any} [icon]
	 * @property {boolean} [small]
	 */

	/** @type {Props} */
	let {
		class: className = '',
		options = [],
		fields = $bindable(defaultFields),
		using = $bindable({ default: Item }),
		value = $bindable(null),
		title = null,
		icon = null,
		small = false,
		...events
	} = $props()

	// $effect.pre(() => {
	// using = { default: Item, ...using }
	// fields = { ...defaultFields, ...fields }
	// })
	let internalUsing = $derived({ default: Item, ...using })
	let internalFields = $derived({ ...defaultFields, ...fields })

	let emitter = createEmitter(events, ['change'])
	let offsetTop = $state(0)
	let open = $state(false)
	let icons = defaultStateIcons.selector

	function handleSelect(data) {
		open = false
		emitter.change(data)
	}
</script>

<!-- svelte-ignore a11y_no_noninteractive_tabindex -->
<drop-down
	class="dropdown relative flex cursor-pointer select-none {className}"
	class:open
	aria-haspopup="true"
	aria-controls="menu"
>
	<button
		class="flex items-center"
		bind:clientHeight={offsetTop}
		tabindex="0"
		use:dismissable
		ondismiss={() => (open = false)}
		onfocus={() => (open = true)}
		onblur={() => (open = false)}
	>
		<span class="flex items-center">
			{#if icon !== null}
				<Icon name={icon} />
			{/if}
			{#if !small && title}
				<p class="flex w-full">{title}</p>
			{/if}
			{#if open}
				<Icon name={icons.opened} />
			{:else}
				<Icon name={icons.closed} />
			{/if}
		</span>
	</button>
	{#if open}
		<Slider top={offsetTop + 4}>
			<List
				items={options}
				fields={internalFields}
				using={internalUsing}
				bind:value
				onselect={handleSelect}
				tabindex="-1"
			/>
		</Slider>
	{/if}
</drop-down>
