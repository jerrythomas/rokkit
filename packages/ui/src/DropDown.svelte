<script>
	import { defaultFields, defaultStateIcons } from '@rokkit/core'
	import { dismissable } from '@rokkit/actions'
	import Icon from './Icon.svelte'
	import Slider from './Slider.svelte'
	import Item from './Item.svelte'
	import List from './List.svelte'

	let {
		class: className,
		options = [],
		fields = defaultFields,
		using = { default: Item },
		value = null,
		title = null,
		icon = null,
		small = false,
		disabled = false
	} = $props()

	using = { default: Item, ...using }
	fields = { ...defaultFields, ...fields }

	let offsetTop = $state(0)
	let open = $state(false)
	let icons = defaultStateIcons.selector

	function handleSelect(event) {
		open = false
		// dispatch('change', event.detail)
	}
</script>

<drop-down
	class="dropdown relative flex cursor-pointer select-none {className}"
	class:open
	aria-haspopup="true"
	aria-controls="menu"
>
	<button
		class="flex items-center"
		bind:clientHeight={offsetTop}
		tabindex="0"
		use:dismissable
		ondismiss={() => (open = false)}
		onfocus={() => (open = true)}
		onblur={() => (open = false)}
	>
		<span class="flex items-center">
			{#if icon !== null}
				<Icon name={icon} />
			{/if}
			{#if !small && title}
				<p class="flex w-full">{title}</p>
			{/if}
			{#if open}
				<Icon name={icons.opened} />
			{:else}
				<Icon name={icons.closed} />
			{/if}
		</span>
	</button>
	{#if open}
		<Slider top={offsetTop + 4}>
			<List items={options} {fields} {using} bind:value onselect={handleSelect} tabindex="-1" />
		</Slider>
	{/if}
</drop-down>
