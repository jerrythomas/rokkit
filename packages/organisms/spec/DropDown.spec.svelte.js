import { describe, expect, it, beforeEach, vi, afterAll } from 'vitest'
import { cleanup, render, fireEvent } from '@testing-library/svelte'
import DropDown from '../src/DropDown.svelte'
import { tick } from 'svelte'
import MockItem from './mocks/MockItem.svelte'
import 'validators/mocks'

describe('DropDown.svelte', () => {
	const events = ['change']
	const handlers = {}

	beforeEach(() => {
		cleanup()
		events.forEach((e) => (handlers[e] = vi.fn()))
	})
	afterAll(() => vi.resetAllMocks())

	it('should render empty text', () => {
		const { container } = render(DropDown, {
			props: { options: [], fields: {} }
		})
		const button = container.querySelector('drop-down > button')
		expect(button.textContent.trim()).toBe('')
	})

	it('should render using icon only', async () => {
		const { container } = render(DropDown, {
			props: { options: ['One', 'Two', 'Three'], fields: {}, icon: 'theme' }
		})
		const button = container.querySelector('drop-down > button')
		expect(button).toMatchSnapshot()
		await fireEvent.click(button)
		await tick()
		expect(container).toMatchSnapshot()
	})

	it('should render title', async () => {
		const props = $state({ options: [], fields: {}, title: 'Select a value' })
		const { container } = render(DropDown, { props })

		const button = container.querySelector('drop-down > button')
		expect(button.querySelector('p').textContent.trim()).toBe('Select a value')

		props.icon = 'theme'
		await tick()
		expect(button).toMatchSnapshot()
		// expect(button.querySelector('icon i').getAttribute('class')).toBe('theme')

		props.icon = 'other'
		await tick()
		expect(button).toMatchSnapshot()

		props.title = 'Select a different value'
		await tick()
		expect(button.querySelector('p').textContent.trim()).toBe('Select a different value')
	})

	it('should toggle open state on focus change', async () => {
		const { container } = render(DropDown, {
			props: { options: [], fields: {} }
		})

		expect(container).toMatchSnapshot()
		const button = container.querySelector('drop-down > button')

		await fireEvent.focus(button)
		expect(button.parentNode.classList.contains('open')).toBe(true)
		await fireEvent.blur(button)
		expect(button.parentNode.classList.contains('open')).toBe(false)
	})

	it('should dispatch a change event when an item is selected', async () => {
		let selected = null
		const options = [{ text: 'foo' }, { text: 'bar' }]
		const props = $state({
			options,
			onchange: handlers.change,
			value: null
		})
		const { container } = render(DropDown, { props })

		const button = container.querySelector('drop-down > button')
		await fireEvent.focus(button)
		expect(button.parentNode.classList.contains('open')).toBe(true)

		container.querySelectorAll('drop-down list item').forEach(async (item, index) => {
			selected = { item: options[index], indices: [index] }
			await fireEvent.click(item)
			// expect(props).value).toEqual(selected.item)

			expect(handlers.change).toHaveBeenCalled()
			expect(handlers.change).toHaveBeenCalledWith(selected)
		})
	})

	it('should render string array', async () => {
		const props = $state({
			options: ['One', 'Two', 'Three'],
			fields: {},
			onchange: handlers.change
		})
		const { container } = render(DropDown, { props })
		expect(container).toBeTruthy()

		const wrapper = container.querySelector('drop-down')
		const button = container.querySelector('drop-down > button')
		await fireEvent.focus(button)
		await tick()

		expect(container).toMatchSnapshot()
		const items = container.querySelectorAll('drop-down list item')
		expect(items.length).toEqual(3)
		await fireEvent.click(items[1])
		await tick()

		// expect(handlers.change).toHaveBeenCalledWith({
		// 	item: 'b',
		// 	indices: [1]
		// })
		expect(wrapper.classList.contains('open')).toBe(false)
		await fireEvent.blur(button)

		props.options = ['x', 'y', 'z']
		props.icon = 'theme'
		// setProperties(component, { options: ['x', 'y', 'z'], icon: 'theme' })
		await fireEvent.focus(button)
		await tick()
		expect(container).toMatchSnapshot()
	})

	it('should render items using field mapping', async () => {
		const props = $state({
			options: [{ text: 1 }, { text: 2 }, { text: 3 }]
		})
		const { container } = render(DropDown, { props })
		expect(container).toBeTruthy()
		await fireEvent.click(container.querySelector('drop-down '))
		await tick()
		expect(container).toMatchSnapshot()
	})

	it('should render items using custom mapping', async () => {
		const options = [{ num: 1 }, { num: 2 }, { num: 3 }]
		const props = $state({
			options,
			value: options[1],
			fields: { text: 'num' }
		})

		const { container } = render(DropDown, { props })
		expect(container).toBeTruthy()
		await fireEvent.click(container.querySelector('drop-down '))
		await tick()
		expect(container).toMatchSnapshot()
	})

	it('should render items using default component, when invalid component is provided', async () => {
		const options = [{ num: 1 }, { num: 2 }, { num: 3 }]
		const props = $state({
			options,
			value: options[1],
			fields: { text: 'num', component: 'num' }
		})
		const { container } = render(DropDown, { props })
		expect(container).toBeTruthy()
		await fireEvent.click(container.querySelector('drop-down '))
		await tick()
		expect(container).toMatchSnapshot()
	})

	it('should render with alternate class', async () => {
		const props = $state({
			options: [{ num: 1 }, { num: 2 }, { num: 3 }],
			fields: { text: 'num' },
			class: 'myClass'
		})
		const { container, component } = render(DropDown, { props })
		let classes = Array.from(container.querySelector('drop-down').classList)

		expect(classes).toContain('myClass')
		props.class = 'myClass2'
		await tick()
		classes = Array.from(container.querySelector('drop-down').classList)
		expect(classes).not.toContain('myClass')
		expect(classes).toContain('myClass2')
	})

	it('should render items using custom component', async () => {
		const options = [{ num: 1, component: 'custom' }, { num: 2 }, { num: 3 }]
		const props = $state({
			options,
			fields: { text: 'num' },
			using: { custom: MockItem },
			value: options[0]
		})

		const { container } = render(DropDown, { props })
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
		const item = container.querySelector('drop-down ')
		await fireEvent.click(item)
		await tick()
		expect(container).toMatchSnapshot()

		props.using = { default: MockItem, custom: MockItem }
		await tick()
		expect(container).toMatchSnapshot()
	})

	it('should close drop down on escape key', async () => {
		const options = [{ text: 'a' }, { text: 'b' }, { text: 'c' }]
		const props = $state({
			options,
			value: options[1],
			onchange: handlers.change
		})
		const { container } = render(DropDown, { props })

		const wrapper = container.querySelector('drop-down')
		await fireEvent.focus(wrapper.querySelector('button'))

		await tick()
		let classes = Array.from(wrapper.classList)
		expect(classes).toContain('open')
		expect(wrapper).toMatchSnapshot()

		await fireEvent.keyUp(wrapper, {
			key: 'Escape'
		})
		await tick()
		classes = Array.from(wrapper.classList)
		expect(classes).not.toContain('open')
		expect(wrapper).toMatchSnapshot()
	})

	// it('should handle arrow keys for navigation and select', async () => {
	// 	const options = [{ text: 'a' }, { text: 'b' }, { text: 'c' }]
	// 	const { container, component } = render(DropDown, {
	// 		options: options,
	// 		value: options[1]
	// 	})

	// 	expect(container.querySelector('drop-down ')).toMatchSnapshot()
	// 	component.$on('change', handlers.change)
	// 	const select = container.querySelector('drop-down ')

	// 	await fireEvent.keyDown(select, { key: 'ArrowDown' })
	// 	await tick()
	// 	await fireEvent.keyDown(select, { key: 'ArrowDown' })
	// 	await tick()
	// 	await fireEvent.keyDown(select, { key: 'Enter' })
	// 	await tick()
	// 	expect(handlers.change).toHaveBeenCalled()
	// expect(getPropertyValue(component, 'value')).toEqual(options[2])
	// expect(container.querySelector('drop-down ')).toMatchSnapshot()

	// await fireEvent.keyDown(select, { key: 'ArrowUp' })
	// await tick()
	// await fireEvent.keyDown(select, { key: 'ArrowUp' })
	// await tick()
	// await fireEvent.keyDown(select, { key: 'Enter' })
	// await tick()
	// expect(getPropertyValue(component, 'value')).toEqual(options[1])
	// expect(container.querySelector('drop-down ')).toMatchSnapshot()
	// })
})
