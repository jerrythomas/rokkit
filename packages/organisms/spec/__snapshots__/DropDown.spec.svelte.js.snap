// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DropDown.svelte > should close drop down on escape key 1`] = `
<drop-down
  aria-controls="menu"
  aria-haspopup="true"
  class="dropdown relative flex cursor-pointer select-none open"
>
  <button
    class="flex items-center"
    tabindex="0"
  >
    <span
      class="flex items-center"
    >
      <!---->
       
      <!---->
       
      <rkt-icon
        aria-label="selector-opened"
        class="flex flex-shrink-0 items-center justify-center "
        role="img"
        tabindex="-1"
      >
        <i
          aria-hidden="true"
          class="selector-opened"
        />
      </rkt-icon>
      <!---->
    </span>
  </button>
   
  <scroll
    class="absolute z-10 flex w-full flex-col overflow-scroll menu"
    style="top: 4px;"
  >
    <list
      aria-label="list"
      class="flex w-full flex-shrink-0 select-none flex-col list"
      role="listbox"
      tabindex="-1"
    >
      <!---->
       
      <!---->
      <item
        aria-selected="false"
        class="item"
        data-path="0"
        role="option"
      >
        <!---->
        <!---->
        <!---->
        
        
        <!---->
         
        <p>
          a
        </p>
        <!---->
        <!---->
      </item>
      <item
        aria-selected="true"
        class="item"
        data-path="1"
        role="option"
      >
        <!---->
        <!---->
        <!---->
        
        
        <!---->
         
        <p>
          b
        </p>
        <!---->
        <!---->
      </item>
      <item
        aria-selected="false"
        class="item"
        data-path="2"
        role="option"
      >
        <!---->
        <!---->
        <!---->
        
        
        <!---->
         
        <p>
          c
        </p>
        <!---->
        <!---->
      </item>
      
      <!---->
    </list>
    <!---->
  </scroll>
  <!---->
</drop-down>
`;

exports[`DropDown.svelte > should close drop down on escape key 2`] = `
<drop-down
  aria-controls="menu"
  aria-haspopup="true"
  class="dropdown relative flex cursor-pointer select-none"
>
  <button
    class="flex items-center"
    tabindex="0"
  >
    <span
      class="flex items-center"
    >
      <!---->
       
      <!---->
       
      <rkt-icon
        aria-label="selector-closed"
        class="flex flex-shrink-0 items-center justify-center "
        role="img"
        tabindex="-1"
      >
        <i
          aria-hidden="true"
          class="selector-closed"
        />
      </rkt-icon>
      <!---->
    </span>
  </button>
   
  <scroll
    class="absolute z-10 flex w-full flex-col overflow-scroll menu"
    style="top: 4px;"
  >
    <list
      aria-label="list"
      class="flex w-full flex-shrink-0 select-none flex-col list"
      role="listbox"
      tabindex="-1"
    >
      <!---->
       
      <!---->
      <item
        aria-selected="false"
        class="item"
        data-path="0"
        role="option"
      >
        <!---->
        <!---->
        <!---->
        
        
        <!---->
         
        <p>
          a
        </p>
        <!---->
        <!---->
      </item>
      <item
        aria-selected="true"
        class="item"
        data-path="1"
        role="option"
      >
        <!---->
        <!---->
        <!---->
        
        
        <!---->
         
        <p>
          b
        </p>
        <!---->
        <!---->
      </item>
      <item
        aria-selected="false"
        class="item"
        data-path="2"
        role="option"
      >
        <!---->
        <!---->
        <!---->
        
        
        <!---->
         
        <p>
          c
        </p>
        <!---->
        <!---->
      </item>
      
      <!---->
    </list>
    <!---->
  </scroll>
  <!---->
</drop-down>
`;

exports[`DropDown.svelte > should render items using custom component 1`] = `
<div>
  <drop-down
    aria-controls="menu"
    aria-haspopup="true"
    class="dropdown relative flex cursor-pointer select-none "
  >
    <button
      class="flex items-center"
      tabindex="0"
    >
      <span
        class="flex items-center"
      >
        <!---->
         
        <!---->
         
        <rkt-icon
          aria-label="selector-closed"
          class="flex flex-shrink-0 items-center justify-center "
          role="img"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="selector-closed"
          />
        </rkt-icon>
        <!---->
      </span>
    </button>
     
    <!---->
  </drop-down>
  
</div>
`;

exports[`DropDown.svelte > should render items using custom component 2`] = `
<div>
  <drop-down
    aria-controls="menu"
    aria-haspopup="true"
    class="dropdown relative flex cursor-pointer select-none "
  >
    <button
      class="flex items-center"
      tabindex="0"
    >
      <span
        class="flex items-center"
      >
        <!---->
         
        <!---->
         
        <rkt-icon
          aria-label="selector-closed"
          class="flex flex-shrink-0 items-center justify-center "
          role="img"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="selector-closed"
          />
        </rkt-icon>
        <!---->
      </span>
    </button>
     
    <!---->
  </drop-down>
  
</div>
`;

exports[`DropDown.svelte > should render items using custom component 3`] = `
<div>
  <drop-down
    aria-controls="menu"
    aria-haspopup="true"
    class="dropdown relative flex cursor-pointer select-none "
  >
    <button
      class="flex items-center"
      tabindex="0"
    >
      <span
        class="flex items-center"
      >
        <!---->
         
        <!---->
         
        <rkt-icon
          aria-label="selector-closed"
          class="flex flex-shrink-0 items-center justify-center "
          role="img"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="selector-closed"
          />
        </rkt-icon>
        <!---->
      </span>
    </button>
     
    <!---->
  </drop-down>
  
</div>
`;

exports[`DropDown.svelte > should render items using custom mapping 1`] = `
<div>
  <drop-down
    aria-controls="menu"
    aria-haspopup="true"
    class="dropdown relative flex cursor-pointer select-none "
  >
    <button
      class="flex items-center"
      tabindex="0"
    >
      <span
        class="flex items-center"
      >
        <!---->
         
        <!---->
         
        <rkt-icon
          aria-label="selector-closed"
          class="flex flex-shrink-0 items-center justify-center "
          role="img"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="selector-closed"
          />
        </rkt-icon>
        <!---->
      </span>
    </button>
     
    <!---->
  </drop-down>
  
</div>
`;

exports[`DropDown.svelte > should render items using default component, when invalid component is provided 1`] = `
<div>
  <drop-down
    aria-controls="menu"
    aria-haspopup="true"
    class="dropdown relative flex cursor-pointer select-none "
  >
    <button
      class="flex items-center"
      tabindex="0"
    >
      <span
        class="flex items-center"
      >
        <!---->
         
        <!---->
         
        <rkt-icon
          aria-label="selector-closed"
          class="flex flex-shrink-0 items-center justify-center "
          role="img"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="selector-closed"
          />
        </rkt-icon>
        <!---->
      </span>
    </button>
     
    <!---->
  </drop-down>
  
</div>
`;

exports[`DropDown.svelte > should render items using field mapping 1`] = `
<div>
  <drop-down
    aria-controls="menu"
    aria-haspopup="true"
    class="dropdown relative flex cursor-pointer select-none "
  >
    <button
      class="flex items-center"
      tabindex="0"
    >
      <span
        class="flex items-center"
      >
        <!---->
         
        <!---->
         
        <rkt-icon
          aria-label="selector-closed"
          class="flex flex-shrink-0 items-center justify-center "
          role="img"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="selector-closed"
          />
        </rkt-icon>
        <!---->
      </span>
    </button>
     
    <!---->
  </drop-down>
  
</div>
`;

exports[`DropDown.svelte > should render string array 1`] = `
<div>
  <drop-down
    aria-controls="menu"
    aria-haspopup="true"
    class="dropdown relative flex cursor-pointer select-none open"
  >
    <button
      class="flex items-center"
      tabindex="0"
    >
      <span
        class="flex items-center"
      >
        <!---->
         
        <!---->
         
        <rkt-icon
          aria-label="selector-opened"
          class="flex flex-shrink-0 items-center justify-center "
          role="img"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="selector-opened"
          />
        </rkt-icon>
        <!---->
      </span>
    </button>
     
    <scroll
      class="absolute z-10 flex w-full flex-col overflow-scroll menu"
      style="top: 4px;"
    >
      <list
        aria-label="list"
        class="flex w-full flex-shrink-0 select-none flex-col list"
        role="listbox"
        tabindex="-1"
      >
        <!---->
         
        <!---->
        <item
          aria-selected="false"
          class="item"
          data-path="0"
          role="option"
        >
          <!---->
          <!---->
           
          <p>
            One
          </p>
          <!---->
          <!---->
        </item>
        <item
          aria-selected="false"
          class="item"
          data-path="1"
          role="option"
        >
          <!---->
          <!---->
           
          <p>
            Two
          </p>
          <!---->
          <!---->
        </item>
        <item
          aria-selected="false"
          class="item"
          data-path="2"
          role="option"
        >
          <!---->
          <!---->
           
          <p>
            Three
          </p>
          <!---->
          <!---->
        </item>
        
        <!---->
      </list>
      <!---->
    </scroll>
    <!---->
  </drop-down>
  
</div>
`;

exports[`DropDown.svelte > should render string array 2`] = `
<div>
  <drop-down
    aria-controls="menu"
    aria-haspopup="true"
    class="dropdown relative flex cursor-pointer select-none open"
  >
    <button
      class="flex items-center"
      tabindex="0"
    >
      <span
        class="flex items-center"
      >
        <rkt-icon
          aria-label="theme"
          class="flex flex-shrink-0 items-center justify-center "
          role="img"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="theme"
          />
        </rkt-icon>
        <!---->
         
        <!---->
         
        <rkt-icon
          aria-label="selector-opened"
          class="flex flex-shrink-0 items-center justify-center "
          role="img"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="selector-opened"
          />
        </rkt-icon>
        <!---->
      </span>
    </button>
     
    <scroll
      class="absolute z-10 flex w-full flex-col overflow-scroll menu"
      style="top: 4px;"
    >
      <list
        aria-label="list"
        class="flex w-full flex-shrink-0 select-none flex-col list"
        role="listbox"
        tabindex="-1"
      >
        <!---->
         
        <!---->
        <item
          aria-selected="false"
          class="item"
          data-path="0"
          role="option"
        >
          <!---->
          <!---->
           
          <p>
            x
          </p>
          <!---->
          <!---->
        </item>
        <item
          aria-selected="false"
          class="item"
          data-path="1"
          role="option"
        >
          <!---->
          <!---->
           
          <p>
            y
          </p>
          <!---->
          <!---->
        </item>
        <item
          aria-selected="false"
          class="item"
          data-path="2"
          role="option"
        >
          <!---->
          <!---->
           
          <p>
            z
          </p>
          <!---->
          <!---->
        </item>
        
        <!---->
      </list>
      <!---->
    </scroll>
    <!---->
  </drop-down>
  
</div>
`;

exports[`DropDown.svelte > should render title 1`] = `
<button
  class="flex items-center"
  tabindex="0"
>
  <span
    class="flex items-center"
  >
    <rkt-icon
      aria-label="theme"
      class="flex flex-shrink-0 items-center justify-center "
      role="img"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="theme"
      />
    </rkt-icon>
    <!---->
     
    <p
      class="flex w-full"
    >
      Select a value
    </p>
    <!---->
     
    <rkt-icon
      aria-label="selector-closed"
      class="flex flex-shrink-0 items-center justify-center "
      role="img"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="selector-closed"
      />
    </rkt-icon>
    <!---->
  </span>
</button>
`;

exports[`DropDown.svelte > should render title 2`] = `
<button
  class="flex items-center"
  tabindex="0"
>
  <span
    class="flex items-center"
  >
    <rkt-icon
      aria-label="theme"
      class="flex flex-shrink-0 items-center justify-center "
      role="img"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="other"
      />
    </rkt-icon>
    <!---->
     
    <p
      class="flex w-full"
    >
      Select a value
    </p>
    <!---->
     
    <rkt-icon
      aria-label="selector-closed"
      class="flex flex-shrink-0 items-center justify-center "
      role="img"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="selector-closed"
      />
    </rkt-icon>
    <!---->
  </span>
</button>
`;

exports[`DropDown.svelte > should render using icon only 1`] = `
<button
  class="flex items-center"
  tabindex="0"
>
  <span
    class="flex items-center"
  >
    <rkt-icon
      aria-label="theme"
      class="flex flex-shrink-0 items-center justify-center "
      role="img"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="theme"
      />
    </rkt-icon>
    <!---->
     
    <!---->
     
    <rkt-icon
      aria-label="selector-closed"
      class="flex flex-shrink-0 items-center justify-center "
      role="img"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="selector-closed"
      />
    </rkt-icon>
    <!---->
  </span>
</button>
`;

exports[`DropDown.svelte > should render using icon only 2`] = `
<div>
  <drop-down
    aria-controls="menu"
    aria-haspopup="true"
    class="dropdown relative flex cursor-pointer select-none "
  >
    <button
      class="flex items-center"
      tabindex="0"
    >
      <span
        class="flex items-center"
      >
        <rkt-icon
          aria-label="theme"
          class="flex flex-shrink-0 items-center justify-center "
          role="img"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="theme"
          />
        </rkt-icon>
        <!---->
         
        <!---->
         
        <rkt-icon
          aria-label="selector-closed"
          class="flex flex-shrink-0 items-center justify-center "
          role="img"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="selector-closed"
          />
        </rkt-icon>
        <!---->
      </span>
    </button>
     
    <!---->
  </drop-down>
  
</div>
`;

exports[`DropDown.svelte > should toggle open state on focus change 1`] = `
<div>
  <drop-down
    aria-controls="menu"
    aria-haspopup="true"
    class="dropdown relative flex cursor-pointer select-none "
  >
    <button
      class="flex items-center"
      tabindex="0"
    >
      <span
        class="flex items-center"
      >
        <!---->
         
        <!---->
         
        <rkt-icon
          aria-label="selector-closed"
          class="flex flex-shrink-0 items-center justify-center "
          role="img"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="selector-closed"
          />
        </rkt-icon>
        <!---->
      </span>
    </button>
     
    <!---->
  </drop-down>
  
</div>
`;
